File: pdu/python/hako_srv_msgs/pdu_conv_AddTwoIntsRequestPacket.py

import struct
from .pdu_pytype_AddTwoIntsRequestPacket import AddTwoIntsRequestPacket
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class
from ..hako_srv_msgs.pdu_conv_ServiceRequestHeader import *
from ..hako_srv_msgs.pdu_conv_AddTwoIntsRequest import *



def pdu_to_py_AddTwoIntsRequestPacket(binary_data: bytes) -> AddTwoIntsRequestPacket:
    py_obj = AddTwoIntsRequestPacket()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_AddTwoIntsRequestPacket(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_AddTwoIntsRequestPacket(meta: binary_io.PduMetaData, binary_data: bytes, py_obj: AddTwoIntsRequestPacket, base_off: int):
    # array_type: single 
    # data_type: struct 
    # member_name: header 
    # type_name: ServiceRequestHeader 
    # offset: 0 size: 268 
    # array_len: 1

    tmp_py_obj = ServiceRequestHeader()
    binary_read_recursive_ServiceRequestHeader(meta, binary_data, tmp_py_obj, base_off + 0)
    py_obj.header = tmp_py_obj
    
    # array_type: single 
    # data_type: struct 
    # member_name: body 
    # type_name: AddTwoIntsRequest 
    # offset: 272 size: 16 
    # array_len: 1

    tmp_py_obj = AddTwoIntsRequest()
    binary_read_recursive_AddTwoIntsRequest(meta, binary_data, tmp_py_obj, base_off + 272)
    py_obj.body = tmp_py_obj
    
    return py_obj

File: pdu/javascript/hako_srv_msgs/pdu_conv_AddTwoIntsRequestPacket.js
import * as PduUtils from '../pdu_utils.js';
import { AddTwoIntsRequestPacket } from './pdu_jstype_AddTwoIntsRequestPacket.js';
import { ServiceRequestHeader } from './pdu_jstype_ServiceRequestHeader.js';
import { binary_read_recursive_ServiceRequestHeader, binary_write_recursive_ServiceRequestHeader } from './pdu_conv_ServiceRequestHeader.js';
import { AddTwoIntsRequest } from './pdu_jstype_AddTwoIntsRequest.js';
import { binary_read_recursive_AddTwoIntsRequest, binary_write_recursive_AddTwoIntsRequest } from './pdu_conv_AddTwoIntsRequest.js';


/**
 * Deserializes a binary PDU into a AddTwoIntsRequestPacket object.
 * @param {ArrayBuffer} binary_data
 * @returns { AddTwoIntsRequestPacket }
 */
export function pduToJs_AddTwoIntsRequestPacket(binary_data) {
    const js_obj = new AddTwoIntsRequestPacket();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_AddTwoIntsRequestPacket(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_AddTwoIntsRequestPacket(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: header, type: ServiceRequestHeader (struct)

    {
        const tmp_obj = new ServiceRequestHeader();
        binary_read_recursive_ServiceRequestHeader(meta, binary_data, tmp_obj, base_off + 0);
        js_obj.header = tmp_obj;
    }
    
    // member: body, type: AddTwoIntsRequest (struct)

    {
        const tmp_obj = new AddTwoIntsRequest();
        binary_read_recursive_AddTwoIntsRequest(meta, binary_data, tmp_obj, base_off + 272);
        js_obj.body = tmp_obj;
    }
    
    return js_obj;
}

/**
 * Serializes a AddTwoIntsRequestPacket object into a binary PDU.
 * @param { AddTwoIntsRequestPacket } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_AddTwoIntsRequestPacket(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_AddTwoIntsRequestPacket(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_AddTwoIntsRequestPacket(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: header, type: ServiceRequestHeader (struct)

    {
        binary_write_recursive_ServiceRequestHeader(parent_off + 0, bw_container, allocator, js_obj.header);
    }
    
    // member: body, type: AddTwoIntsRequest (struct)

    {
        binary_write_recursive_AddTwoIntsRequest(parent_off + 272, bw_container, allocator, js_obj.body);
    }
    
}
